test/test_rate_limiter.js

```diff
@@ -69,18 +69,71 @@ describe('Rate limiter', function() {
 
     queue.on(
       'completed',
+      // after every job has been completed
       _.after(numJobs, function() {
         try {
-          expect(new Date().getTime() - startTime).to.be.above(
-            (numJobs - 1) * 1000
-          );
+          var timeDiff = new Date().getTime() - startTime;
+          expect(timeDiff).to.be.above((numJobs - 1) * 1000);
           done();
         } catch (e) {
-          done(e);
+          assert.fail(e);
         }
       })
     );
 
-    queue.on('failed', done);
+    queue.on('failed', function(e) {
+      assert.fail(e);
+    });
+  });
+
+  it('should put a job into the delayed queue when limit is hit', function() {
+    queue.on('failed', function(e) {
+      assert.fail(e);
+    });
+
+    return Promise.all(
+      [1, 2, 3, 4].map(function() {
+        return queue.add({});
+      })
+    ).then(function() {
+      Promise.all(
+        [1, 2, 3, 4].map(function() {
+          return queue.getNextJob();
+        })
+      ).then(function() {
+        return queue.getDelayedCount().then(function(delayedCount) {
+          expect(delayedCount).to.eq(3);
+        });
+      });
+    });
+  });
+
+  it('should not put a job into the delayed queue when discard is true', function() {
+    var newQueue = utils.buildQueue('test rate limiter', {
+      limiter: {
+        max: 1,
+        duration: 1000,
+        discard: true
+      }
+    });
+
+    newQueue.on('failed', function(e) {
+      assert.fail(e);
+    });
+    return Promise.all(
+      [1, 2, 3, 4].map(function() {
+        return newQueue.add({});
+      })
+    ).then(function() {
+      Promise.all(
+        [1, 2, 3, 4].map(function() {
+          return newQueue.getNextJob();
+        })
+      ).then(function() {
+        return newQueue.getDelayedCount().then(function(delayedCount) {
+          expect(delayedCount).to.eq(0);
```

Comment:
Should also do an assert on the main queue to make sure 3 objects are there. Might also want to assert the ordering of the jobs to make sure that is untouched?